driver/pace/driver/grid.py

```diff
@@ -196,6 +199,111 @@ def get_grid(
         return damping_coefficients, driver_grid_data, grid_data
 
 
+@GridInitializerSelector.register("external")
+@dataclasses.dataclass
+class ExternalGridConfig(GridInitializer):
+    """
+    Configuration for grid initialized from external data.
+    Input is from tile files generated by FRE-NCtools methods
+    In its get_grid method it calls the MetricTerms class method
+    from_generated which generates an object of MetricTerms to be
+    used to generate the damping_coefficients, driver_grid_data,
+    and grid_data variables
+    """
+
+    grid_type: Optional[int] = 0
+    grid_file_path: str = "/input/tilefile"
+
+    # TODO: Area read in?
+
+    def get_grid(
+        self,
+        quantity_factory: QuantityFactory,
+        communicator: Communicator,
+    ) -> Tuple[DampingCoefficients, DriverGridData, GridData]:
+
+        pace_log.info("Using external grid data")
+
+        # ToDo: refactor when grid_type is an enum
+        if self.grid_type <= 3:
+            tile_num = (
+                pace.util.get_tile_index(
+                    communicator.rank, communicator.partitioner.total_ranks
+                )
+                + 1
+            )
+            tile_file = self.grid_file_path + str(tile_num) + ".nc"
+        else:
+            tile_file = self.grid_file_path
+
+        ds = xr.open_dataset(tile_file)
+        lon = ds.x.values
+        lat = ds.y.values
+        dx = ds.dx.values
+        dy = ds.dy.values
+        area = ds.area.values
+        nx = ds.nx.values.size
+        ny = ds.ny.values.size
+        npx = ds.nxp.values.size
+        npy = ds.nyp.values.size
+
+        subtile_slice_grid = communicator.partitioner.tile.subtile_slice(
+            rank=communicator.rank,
+            global_dims=[pace.util.X_INTERFACE_DIM, pace.util.Y_INTERFACE_DIM],
+            global_extent=(npx, npy),
+            overlap=True,
+        )
+
+        subtile_slice_dx = communicator.partitioner.tile.subtile_slice(
+            rank=communicator.rank,
+            global_dims=[pace.util.Y_INTERFACE_DIM, pace.util.X_DIM],
+            global_extent=(npy, nx),
+            overlap=True,
+        )
+
+        subtile_slice_dy = communicator.partitioner.tile.subtile_slice(
+            rank=communicator.rank,
+            global_dims=[pace.util.Y_DIM, pace.util.X_INTERFACE_DIM],
+            global_extent=(ny, npx),
+            overlap=True,
+        )
+
+        subtile_slice_area = communicator.partitioner.tile.subtile_slice(
+            rank=communicator.rank,
+            global_dims=[pace.util.Y_DIM, pace.util.X_DIM],
+            global_extent=(ny, nx),
+            overlap=True,
+        )
+
+        metric_terms = MetricTerms.from_external(
+            x=lon[subtile_slice_grid],
+            y=lat[subtile_slice_grid],
+            dx=dx[subtile_slice_dx],
+            dy=dy[subtile_slice_dy],
+            area=area[subtile_slice_area],
```

Comment:
Yes, I figured since it was there to read it in as well, unless this is ill-advised

Reply:
it's fine, just double-checking 👍

Reply:
IIRC, the fortran version merely read in the lat-lon pairs and calculated all other metrics.  It would be valuable to know how the values in the files compare to those calculated by the model code from the lat-lon pairs.

Reply:
We could switch so that Pace behaves the same way and use dx, dy, and area for the tests

Reply:
We did discuss with Lucas whether we should carry the metrics in memory or recompute in the few instances where they are necessary.  I know many of the metrics terms are for exclusively for remapping the winds amongst the various representations of A-, C-, and D-grid.

Reply:
So just reviewed the Fortran code and here's what I got for a few of the cases we care about...

cartesian grid
   * grid_type = 4
   * internally created

orthogonal grid
  * grid_type = 5
  * read in x, y, dx, dy, area

CS grid
  * grid_type < 0
  * read in x, y
  * calculate dx, dy, area

Reply:
Many of the metrics are infrequently used or in specific situations that are not always necessary. I think it would be best for some architectures (especially GPUs) to compute on-the-fly, although for traditional CPUs it may still be best to compute metrics once and save them in memory.

Reply:
My 2 cent on an advanced system for this kind of "infrequent-but-vital" field.

Removing memory pressure on the GPU will quickly become a subject. Since the DSL generates from IR, in all backends, we can technically inject code. Therefore, we could have a system that registers this kind of calculation and at JIT compilation time, when we known about the size of the domain and can do memory pressure projection (which already exists in DaCe backend), we could decide to flip the read from those field by the analytics.

Of course this has a few caveats, especially that it can be prohibitive if the analytic solution calls on other fields that themselves can be inlined.

Ultimately, this kind of decision (save or recompute) should be taken of the hands of the modelers and push down the stack.